File: IT/Containers/Esercizi/C++/esercizio24.cpp
/*
file esercizio24.cpp
compilare con: g++ -std=c++11 esercizio24.cpp
*/

#include <iostream>
#include <iterator>
#include <algorithm>
#include <set>

using std::cout;
using std::endl;
using std::ostream_iterator;
using std::set;

int main()
{
	set<int> a = {1, 2, 3};
	set<int> b = {3, 4, 5};
	set<int> c = {-1, 0, 1};

	cout << "A = [";
	copy(a.begin(), a.end(), ostream_iterator<int>(cout, ", "));
	cout << "\b\b]" << endl;

	cout << "B = [";
	copy(b.begin(), b.end(), ostream_iterator<int>(cout, ", "));
	cout << "\b\b]" << endl;

	cout << "C = [";
	copy(c.begin(), c.end(), ostream_iterator<int>(cout, ", "));
	cout << "\b\b]" << endl;

	/* calcola la differenza tra A e B: tutti gli elementi di A non presenti in B */

	set<int> diff_ab;
	set_difference(a.begin(), a.end(), b.begin(), b.end(), inserter(diff_ab, diff_ab.begin()));

	cout << "Differenza tra A e B: [";
	copy(diff_ab.begin(), diff_ab.end(), ostream_iterator<int>(cout, ", "));
	cout << "\b\b]" << endl;

	/* calcola la differenza tra (A - B) e C: tutti gli elementi di A non presenti in B nè in C */

	set<int> diff_ab_c;
	set_difference(diff_ab.begin(), diff_ab.end(), c.begin(), c.end(), inserter(diff_ab_c, diff_ab_c.begin()));

	cout << "Differenza tra (A - B) e C: [";
	copy(diff_ab_c.begin(), diff_ab_c.end(), ostream_iterator<int>(cout, ", "));
	cout << "\b\b]" << endl;
}

File: IT/1/Containers/Esercizi/Java/Esercizio24.java
/* file Esercizio24.java */

import java.util.SortedSet;
import java.util.TreeSet;

public class Esercizio24 {
	public static void main(String [] args) {
		SortedSet<Integer> a = new TreeSet<Integer>();

		a.add(1);
		a.add(2);
		a.add(3);

		SortedSet<Integer> b = new TreeSet<Integer>();

		b.add(3);
		b.add(4);
		b.add(5);

		SortedSet<Integer> c = new TreeSet<Integer>();

		c.add(-1);
		c.add(0);
		c.add(1);

		System.out.println("A = " + a);
		System.out.println("B = " + b);
		System.out.println("C = " + c);

		/* calcola la differenza tra A e B: tutti gli elementi di A non presenti in B */

		a.removeAll(b);
		System.out.println("Differenza tra A e B: " + a);

		/* calcola la differenza tra (A - B) e C: tutti gli elementi di A non presenti in B nè in C */

		a.removeAll(c);
		System.out.println("Differenza tra (A - B) e C: " + a);
	}
}

